src/Clay/Grid.hs

```diff
@@ -1,58 +1,271 @@
+
+-------------------------------------------------------------------------------
+
+gridRowStart :: GridLocation -> Css
+gridRowStart = key "grid-row-start"
+
+gridRowEnd :: GridLocation -> Css
+gridRowEnd = key "grid-row-end"
+
+gridColumnStart :: GridLocation -> Css
+gridColumnStart = key "grid-column-start"
+
+gridColumnEnd :: GridLocation -> Css
+gridColumnEnd = key "grid-column-end"
+
+gridLocation :: IsSpan -> These Integer GridArea -> GridLocation
+gridLocation isSpan = GridLocation_Data . GridLocationData isSpan
+
+data IsSpan = Span | NoSpan
+  deriving (Show, Eq)
+
+data GridLocation
+  = GridLocation_Keyword Value
+  | GridLocation_Data GridLocationData
+
+instance Val GridLocation where
+  value (GridLocation_Keyword v) = v
+  value (GridLocation_Data d) = value d
+
+instance Auto    GridLocation where auto    = GridLocation_Keyword auto
+instance Inherit GridLocation where inherit = GridLocation_Keyword inherit
+instance Initial GridLocation where initial = GridLocation_Keyword initial
+instance Unset   GridLocation where unset   = GridLocation_Keyword unset
+
+-- See under syntax:
+-- https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column-start
+-- either grid index and/or named grid area is required, but span is optional
+data GridLocationData = GridLocationData IsSpan (These Integer GridArea)
+
+instance Val GridLocationData where
+  value (GridLocationData isSpan indexAndOrGridArea) =
+    if isSpan == Span
+    then value (span :: Value, indexAndOrGridArea)
+    else value indexAndOrGridArea
+
+pattern GridIndex :: Integer -> GridLocation
+pattern GridIndex n = GridLocation_Data (GridLocationData NoSpan (This n))
+
+instance Num GridLocation where
+  -- for index literals
+  fromInteger = GridIndex
+  -- for negative index literals
+  negate (GridIndex index) = GridIndex $ negate index
+  -- in general we don't support arithmetic on this type
+  negate _ = error "negate not defined for GridLocation"
+  abs = error "abs not defined for GridLocation"
+  signum = error "abs not defined for GridLocation"
+  (+) = error "addition not defined for GridLocation"
+  (*) = error "multiplication not defined for GridLocation"
+
+-------------------------------------------------------------------------------
+
+-- | Property defines the template for grid layout
+gridTemplateAreas :: GridTemplateAreas -> Css
+gridTemplateAreas = key "grid-template-areas"
+
+newtype GridTemplateAreas = GridTemplateAreas Value
+  deriving (Val, None, Inherit, Initial, Unset)
+
+instance IsList GridTemplateAreas where
+  type Item GridTemplateAreas = [GridArea]
+  toList = error "toList GridTemplateAreas is not defined"
+  fromList = GridTemplateAreas . value . fromList'
+    where
+      fromList' :: [Item GridTemplateNamedAreas] -> GridTemplateNamedAreas
+      fromList' = fromList
+
+-- have to create a newtype to override the Val instance for lists
+newtype GridTemplateNamedAreas = GridTemplateNamedAreas { unGridTemplateNamedAreas :: [[GridArea]] }
```

Comment:
In this particular case, `toList` should be definable. Even derivable: Can't you just use https://hackage.haskell.org/package/derive-monoid and derive `IsList`?

Reply:
`toList` is defined for this. 
I could derive IsList automatically, but then I wouldn't validate the shape of the grid template.
If you generate a grid a non rectangular grid template then the browser will silently ignore that piece of css.
Crashing at generation of css seemed like a better experience.

I did consider using type level finite vectors to check the shape at compile time, but opted for runtime validation to keep error message simple.

Reply:
> Crashing at generation of css seemed like a better experience.

Crashing at runtime is generally a bad idea for a Haskell library. Clay is not necessarily used only to generate CSS, but also used in Haskell webapps to generate it in *real-time*.

Reply:
@srid I tend to agree. Although it will typically "reliably crash" on startup, so you'll probably notice on deployment. But for Haskell, which I associate with "if it compiles it works", this is also not acceptable in my opinion. Maybe in the long run, Template Haskell should be used for such cases, and then the crash is again on compile time.